File: 68.text-justification.go
package leetcode

import "strings"

/*
 * @lc app=leetcode id=68 lang=golang
 *
 * [68] Text Justification
 */

// @lc code=start
func FullJustifyLine(words [][]rune, width, lineLen int) string {
	spaces := width - lineLen
	numGaps := len(words) - 1
	if numGaps == 0{
		return LeftJustifyLine(words, width)
	}
	gapsWithExtra := spaces % numGaps
	smallGap := strings.Repeat(" ", spaces/numGaps + 1)
	sb := strings.Builder{}
	for i, runes := range words {
		sb.WriteString(string(runes))
		if i < numGaps {
			sb.WriteString(smallGap)
			if i < gapsWithExtra {
				sb.WriteString(" ")
			}
		}
	}
	return sb.String()
}
func LeftJustifyLine(words [][]rune, width int) string {
	sb := strings.Builder{}
	n := len(words)
	for i, runes := range words {
		sb.WriteString(string(runes))
		if i+1 < n {
			sb.WriteString(" ")
		}
	}
	sb.WriteString(strings.Repeat(" ", width - sb.Len()))
	return sb.String()

}
func stringsToRunes(words []string) [][]rune {
	result := make([][]rune, len(words))
	for i, word := range words {
		result[i] =  []rune(word)
	}
	return result
}
func fullJustify(words []string, maxWidth int) []string {
	wordRunes := stringsToRunes(words)
	result := []string{}
	i := 0
	n := len(words)
	for i < n {
		lineLen := len(wordRunes[i])
		j := i + 1

		//How many words fit on this line
		for j < n && lineLen+1+len(wordRunes[j]) <= maxWidth {
			lineLen += 1 + len(wordRunes[j])
			j++
		}
		//the final line and one word lines should be left justified
		if j == n || j-i == 1 {
			result = append(result, LeftJustifyLine(wordRunes[i:j], maxWidth))
		} else {
			//All other lines should be fully_justified
			result = append(result, FullJustifyLine(wordRunes[i:j], maxWidth, lineLen))
		}
		//Move onto the next line
		i = j

	}
	return result
}

// @lc code=end

File: 68.text-justification.py
#
# @lc app=leetcode id=68 lang=python3
#
# [68] Text Justification
#

# @lc code=start
def required_spaces(words):
    """ 
    How many space is required to make a sentence out of the words.
    Every two words needs atelast one space between them.
    """
    return max(len(words) -1, 0)

def left_justify_line(words, width):
    line = " ".join(words)
    return f"{line:{' '}<{width}}"
    
def full_justify_line(words, width, num_chars=None):
    """
    Pads the words with spaces so that they reach the given width 
    the padding is apllied as evenly as possible between the words.
    i.e. no two gaps between words ever differ by more than 1
    """
    if not None:
        #optionallly you can supply the total number of chars in the words
        #if not supplied wi simply compute them
        num_chars = sum(map(len, words))
    
    spaces = width - num_chars
    if spaces < required_spaces(words):
        raise ValueError(f"The number of chars exceeds the width can't add negative padding")
    
    spaces_per_gap, gaps_with_extra = divmod(spaces, len(words) -1)
    small_gap = " "*spaces_per_gap
    if gaps_with_extra == 0:
        return small_gap.join(words)
    big_gap = small_gap + " "
    return big_gap.join(words[:gaps_with_extra]) +big_gap+small_gap.join(words[gaps_with_extra:])

class Solution:
    def fullJustify(self, words: List[str], maxWidth: int) -> List[str]:
        #The extra spaces need to be evenly divided among the wrods on a line
        result = []
        words.reverse()
        cur_line = []
        num_chars = 0
        while words:
            cur_word = words.pop()
            word_len = len(cur_word)
            #The current word can't fit into this line
            if num_chars + required_spaces(cur_line) + word_len >= maxWidth :
                #We have to start a new line 
                #First add the curent line to the result, don't add empty lines
                if len(cur_line) == 1:
                    result.append(left_justify_line(cur_line, maxWidth))
                elif len(cur_line) > 1:
                    result.append(full_justify_line(cur_line, maxWidth))
                    
                #Second start the new line with the current word
                cur_line.clear()
                num_chars = 0
            
            if word_len == maxWidth:
                result.append(cur_word)
                num_chars = 0
                continue

            cur_line.append(cur_word)
            num_chars += word_len
    
    
        if len(cur_line) > 0:
            result.append(left_justify_line(cur_line, maxWidth))
        return result
        
# @lc code=end

